add Triangulation::ghost_owners()

tjhei opened the pull request.

gassmoeller commented:
Very nice feature. Was this motivated by aspect pr 411? I made a preliminary implementation there (very inefficient), but this can nicely replace my own.

tjhei commented:
No, I need this for something else but even better. :-)

davydden commented on include/deal.II/distributed/tria_base.h:
```diff
@@ -132,6 +132,13 @@ namespace parallel
+     * list of processor @p j, then @p j will also be contained in the list of
+     * processor @p i.
+     **/
+    virtual const std::vector<unsigned int> &ghost_neighbors () const;
```
i wonder if `ghost_owners` is a more descriptive name? I think that what you do: if cell is ghost from the perspective of the current processor, put it's subdomain id (== mpi rank) to the list.


> ...at least one ghost cell adjacent to our cells

I may be wrong, but i read that line as if these cells are ghosts for other processors,
whereas in reality these ghost cells are ghost only for the current processor. 
Perhaps the description could be slightly changed to reflect it.

tjhei replied:
good suggestion, I will update.

bangerth commented on doc/news/changes.h:
```diff
@@ -159,6 +159,12 @@ inconvenience this causes.
 <a name="general"></a>
 <h3>General</h3>
 <ol>
+  <li> New: Triangulation::ghost_neighbors() returns a list of MPI ranks of owners
```
Well, it isn't a list but a vector. But that made me think about the fact that what you probably *really* want is a `std::set`.

tjhei replied:
The data is a list (or array if you will) of unique ranks (so a set in mathematical terms), which I happen to return in a std::vector (not in a std::list of course!) because I think that is the most flexible data structure to use. I thought about using a std::set here and I am willing to change it if you prefer. I generally prefer std::vector over std::set because it is more efficient/faster, compact, I can memcpy data out, etc..
std::set or std::vector?

bangerth replied:
I like std::set better. It automatically answers the question of duplicates and is sorted for when you need that.

As for `memcpy`: I prefer `std::copy` anyway :-)

bangerth commented on include/deal.II/distributed/tria_base.h:
```diff
@@ -160,6 +167,7 @@ namespace parallel
       std::vector<unsigned int> n_locally_owned_active_cells;
       types::global_dof_index   n_global_active_cells;
       unsigned int              n_global_levels;
+      std::vector<unsigned int> ghost_neighbors;
```
I know the other members aren't documented either, but could you add something for the new variable at least?

tjhei commented:
okay, ready for second review (I also added ```level_ghost_owners()``` while I am there).

bangerth commented on include/deal.II/distributed/tria_base.h:
```diff
@@ -132,6 +132,26 @@ namespace parallel
+     * @note: If @p i is contained in the list of processor @p j, then @p j
+     * will also be contained in the list of processor @p i.
+     **/
+    virtual const std::set<unsigned int> &level_ghost_owners () const;
```
Depending on where you plan on using this, would it make sense to let this function take a level argument, or return a vector of sets?

tjhei replied:
I only need the union over all levels for now. This way I can do one send/recv loop and handle all levels at once. I don't think it matters much (the neighbors will be very similar if not the same for each level). So the code will be simpler but I might need to communicate with a few more processors.

bangerth replied:
OK if you've thought this through. I was thinking of cases where one level uses a (geometrically) completely different partition and where the neighbors on one level have nothing to do with the neighbors on another level. But you know much better what you want to do with the function.

tjhei replied:
I admit that I implemented exactly what I currently needed. Do you want me to go with the by level implementation? Or both?

bangerth replied:
I don't care. You know what your algorithms will require. I simply wanted to point out that with relatively little effort, the function could be generalized if you could foresee a use of this generalization. If you can't, then there is no reason to make it more complicated than necessary.

bangerth commented on include/deal.II/distributed/tria_base.h:
```diff
@@ -132,6 +132,26 @@ namespace parallel
+     * @note: If @p i is contained in the list of processor @p j, then @p j
+     * will also be contained in the list of processor @p i.
+     **/
+    virtual const std::set<unsigned int> &ghost_owners () const;
```
Do these functions have to be virtual, btw?

tjhei replied:
Good question. It is a virtual base class and it doesn't hurt performance-wise. I might need it (see below).

tjhei replied:
The way I solved it now I don't need it, so I can remove it if you want. But it could be that shared::Triangulation needs to override the implementation, because level_subdomain_ids are currently not distributed automatically.

bangerth replied:
If you don't need it now, remove the virtual marker. When I read code, if a function is virtual it always means to me that I will have to find other places where it may be overridden, and so it increases the complexity and interconnectedness of code. If it turns out later that you do need it to be virtual after all, you can always add the marker again.

bangerth commented on include/deal.II/distributed/tria_base.h:
```diff
@@ -157,9 +177,26 @@ namespace parallel
     struct NumberCache
     {
+      /**
+       * number of locally owned cells per MPI rank
```
Can you make these comments complete sentences with a capital letter and a period at the end, while you're there?

bangerth commented on source/distributed/tria_base.cc:
```diff
@@ -163,6 +166,32 @@ namespace parallel
+           cell != this->end();
+           ++cell)
+        if (cell->level_subdomain_id() != numbers::artificial_subdomain_id
+            && cell->level_subdomain_id() != locally_owned_subdomain())
```
Aren't these two conditions equivalent to checking whether the cell is ghost? Or can you only call is_ghost on active cells?

tjhei replied:
1. There is a difference between ```ghost()``` and ```level_ghost()```: ```ghost()``` uses ```subdomain_id```, only works on active cells, and considers neighbors that are on different levels. ```level_ghost()``` uses the ```level_subdomain_id```, is defined on each level, and neighbors only exist on the same level.
2. There is currently no function ```level_ghost()``` so I have to write it out by hand.

bangerth commented on source/distributed/tria_base.cc:
```diff
@@ -163,6 +166,32 @@ namespace parallel
+
+    {
+      // find level ghost owners
+      for (typename Triangulation<dim,spacedim>::cell_iterator
```
I Think you should only run this loop if mgdofs have actually been distributed.

tjhei replied:
Problem: I don't know if this is the case, because we only do this in ```p::d::Triangulation``` and this code currently sits in the base class ```p::Triangulation```. I am moving stuff around...

tjhei commented:
okay, I tried to address your comments, @bangerth .

bangerth commented:
Think about these last two issues. If you're happy with your design, please merge.

tjhei commented:
removed virtual, squashed, and thought about returning owners by level, but I a) don't need it right now (can be added later) and b) the sets will be very similar if not equal.
